File: app/src/main/java/com/example/tetrisapp/util/LeaderboardUtil.java
package com.example.tetrisapp.util;

import android.app.ActivityManager;
import android.util.Log;

import com.example.tetrisapp.data.local.dao.LeaderboardDao;
import com.example.tetrisapp.data.remote.LeaderboardService;
import com.example.tetrisapp.interfaces.Callback;
import com.example.tetrisapp.model.local.entity.LeaderboardEntry;
import com.example.tetrisapp.model.remote.callback.SimpleCallback;
import com.example.tetrisapp.model.remote.request.ScorePayload;
import com.example.tetrisapp.model.remote.request.SubmitScoresPayload;
import com.example.tetrisapp.model.remote.response.DefaultPayload;
import com.example.tetrisapp.model.remote.response.ResponseSubmitScore;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.List;
import java.util.TimeZone;
import java.util.stream.Collectors;

import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.schedulers.Schedulers;

public class LeaderboardUtil {
    public static final String TAG = "LeaderboardUtil";

    private final FirebaseUser user;
    private final LeaderboardDao leaderboardDao;
    private final LeaderboardService leaderboardService;
    private final String token;

    public LeaderboardUtil(LeaderboardDao leaderboardDao) {
        this.leaderboardDao = leaderboardDao;
        this.leaderboardService = null;
        this.token = null;
        this.user = FirebaseAuth.getInstance().getCurrentUser();
    }

    public LeaderboardUtil(String token, LeaderboardDao leaderboardDao, LeaderboardService leaderboardService) {
        this.leaderboardDao = leaderboardDao;
        this.leaderboardService = leaderboardService;
        this.token = token;
        this.user = FirebaseAuth.getInstance().getCurrentUser();
    }

    public void insert(LeaderboardEntry entry) {
        if (user != null) {
            String ISODate = DateTimeUtil.toISOString(entry.date);
            String uid = user.getUid();
            entry.hash = HashUtil.sha256(String.format("%s_%s", uid, ISODate));
        }

        leaderboardDao
                .insert(entry)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        (data) -> {}, throwable -> Log.e(TAG, throwable.getLocalizedMessage()));
    }

    public void synchronise() {
        if (leaderboardService == null && token == null) return;

        generateHashes(() -> leaderboardDao
                .getByUploaded(false)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        (data) -> {
                            if (data.size() == 0) return;
                            
                            List<ScorePayload> scores = data.stream()
                                    .map(entry -> new ScorePayload(entry.score, entry.lines, entry.level, entry.date.getTime()))
                                    .collect(Collectors.toList());

                            leaderboardService
                                    .submitScores(new SubmitScoresPayload(token, scores))
                                    .enqueue(new SimpleCallback.Builder<List<ResponseSubmitScore>>().setOnSuccessCallback((call, res) -> {
                                        res.body().forEach(result -> leaderboardDao.getByHash(result.hash)
                                                .subscribeOn(Schedulers.io())
                                                .observeOn(AndroidSchedulers.mainThread())
                                                .subscribe(entry -> {
                                                    entry.uploaded = result.completed;
                                                    update(entry);
                                                }));
                                    }).build());
                        }, throwable -> Log.e(TAG, throwable.getLocalizedMessage())));
    }

    private void update(LeaderboardEntry entry) {
        leaderboardDao.update(entry).subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(() -> {}, (t) -> {});
    }

    private void generateHashes(Callback callback) {
        if (leaderboardService == null) return;

        leaderboardDao
                .getWithoutHash()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        (data) -> {
                            if (user == null) return;
                            String uid = user.getUid();

                            data.forEach(entry -> {
                                String ISODate = DateTimeUtil.toISOString(entry.date);
                                entry.hash = HashUtil.sha256(String.format("%s_%s", uid, ISODate));
                                update(entry);
                            });

                            callback.call();
                        }, throwable -> {
                            Log.e(TAG, throwable.getLocalizedMessage());
                        });
    }
}

File: app/src/main/java/com/example/tetrisapp/util/LeaderboardUtil.kt
package com.example.tetrisapp.util

import android.util.Log
import com.example.tetrisapp.data.local.dao.LeaderboardDao
import com.example.tetrisapp.model.local.LeaderboardEntry
import com.example.tetrisapp.model.remote.request.ScorePayload
import com.example.tetrisapp.model.remote.response.DefaultPayload
import com.example.tetrisapp.model.remote.response.PersonalRecord
import com.example.tetrisapp.model.remote.response.PublicRecord
import com.example.tetrisapp.model.remote.response.ResponseSubmitScore
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import io.ktor.serialization.gson.*
import kotlinx.coroutines.*

class LeaderboardUtil(
    private val token: String?,
    private val apiRoute: String,
    private val leaderboardDao: LeaderboardDao
) {
    private val user: FirebaseUser? = FirebaseAuth.getInstance().currentUser
    private var call: HttpResponse? = null

    fun insert(entry: LeaderboardEntry) {
        if (user != null) {
            val isoDate = DateTimeUtil.toISOString(entry.date!!)
            val uid = user.uid
            entry.hash = HashUtil.sha256(String.format("%s_%s", uid, isoDate))
        }

        MainScope().launch(Dispatchers.IO) {
            leaderboardDao.insert(entry)
        }
    }

    fun synchronise() {
        if (token == null) return

        generateHashes {
            val coroutineExceptionHandler = CoroutineExceptionHandler{_, t ->
                Log.e(TAG, t.localizedMessage ?: "")
            }

            MainScope().launch(Dispatchers.IO + coroutineExceptionHandler) {
                leaderboardDao.getByUploaded(false)?.let { entries ->
                    if (entries.isEmpty()) return@launch
                    val scores = entries.map { entry ->
                        ScorePayload(
                            entry.score,
                            entry.lines,
                            entry.level,
                            entry.date!!.time
                        )
                    }

                    HttpClient() {
                        install(ContentNegotiation) {
                            gson()
                        }
                    }.use { client ->
                        call = client.post(apiRoute + "leaderboard/submit") {
                            setBody(mapOf(
                                "idToken" to token,
                                "scores" to scores
                            ))
                            contentType(ContentType.Application.Json)
                        }
                        call?.let { response ->
                            val body: DefaultPayload<List<ResponseSubmitScore>> = response.body()

                            if (response.status.isSuccess() && body.success) {
                                body.data?.forEach { score ->
                                    leaderboardDao.getByHash(score.hash)?.let {
                                        it.uploaded = score.completed
                                        leaderboardDao.update(it)
                                    }
                                }
                            } else {
                                throw(Exception(body.message))
                            }
                        }
                    }
                }
            }
        }

        getUploadedScores()
    }

    private fun getUploadedScores() {
        val coroutineExceptionHandler = CoroutineExceptionHandler{_, t ->
            Log.e(TAG, t.localizedMessage ?: "")
        }

        MainScope().launch(Dispatchers.IO + coroutineExceptionHandler) {
            FirebaseAuth.getInstance().currentUser?.uid?.let {
                HttpClient() {
                    install(ContentNegotiation) {
                        gson()
                    }
                }.use { client ->
                    call = client.post(apiRoute + "leaderboard/get/" + it) {
                        setBody(
                            mapOf("idToken" to token)
                        )
                        contentType(ContentType.Application.Json)
                    }
                    call?.let { response ->
                        val body: DefaultPayload<List<PersonalRecord>> = response.body()
                        if (response.status.isSuccess() && body.success) {
                            body.data?.forEach { record ->
                                val entry = LeaderboardEntry()
                                entry.timeInGame = record.timeInGame
                                entry.score  = record.score
                                entry.level = record.level
                                entry.lines = record.lines
                                entry.hash = record.hash
                                entry.date = record.date
                                entry.uploaded = true
                                leaderboardDao.insert(entry)
                            }
                        } else {
                            throw(Exception(body.message))
                        }
                    }
                }
            }
        }
    }

    private fun generateHashes(callback: () -> Unit) {
        MainScope().launch(Dispatchers.IO) {
            leaderboardDao.getWithoutHash()?.let { data ->
                if (user == null) return@launch
                val uid = user.uid
                data.forEach { entry ->
                    val isoDate = DateTimeUtil.toISOString(entry.date!!)
                    entry.hash = HashUtil.sha256(String.format("%s_%s", uid, isoDate))
                    leaderboardDao.update(entry)
                }
                callback()
            }
        }
    }

    companion object {
        const val TAG = "LeaderboardUtil"
    }
}